### PR TITLE
Add caches to SourceData

### PR DESCRIPTION
**Pull request by robertbastian**

### CODE REVIEW

**Comment by sffc** [comment]
Contrasting the approach here, #1859, with the one in #1858:

- Both approaches store a cache of `dyn Any + Send + Sync` trait objects
- #1858 puts the cache in SourceData itself, whereas #1859 puts it nested inside CldrPaths (and presumably UpropsPaths)
- #1858 introduces a new dependency on `moka`, whereas #1859 uses the pre-existing `elsa` dependency
- #1858 has a generalized parsing function, resulting in more flexibility and different Serde impls (uprops needs toml but cldr needs json), whereas #1859 uses a `serde::Deserialize` bound and assumes json, resulting in simpler code at the call site

**Comment by robertbastian** [comment]
What I've done in addition to caching is to route all FS access through `SourceData`, including `ls`es. This way we can implement zip-reading directly in `SourceData` instead of having to touch all the transformers.

**Comment by sffc** [comment]
Two remaining questions versus my draft:

1. `moka` or `elsa` ?
2. Single cache or multiple caches?

**Comment by zbraniecki** [comment]
`elsa` doesn't allow for cache invalidation, am I correct?

**Comment by robertbastian** [comment]
No, but we shouldn't need that. If the user changes the CLDR data while datagen is running I don't think we have to behave deterministically.

**Comment by robertbastian** [comment]
I don't see the advantage of an arc over a shared reference. The transformers only exist in the lifetime of the `SourceData` anyway, so we can just borrow, we don't need to start counting references.

**Comment by sffc** [comment]
The `Arc` is so that the cache can drop data when it gets full.

**Comment by robertbastian** [comment]
Well we can swap this out later if the cache grows too large. This PR makes all transformers work on references.

File: provider/core/src/error.rs [hunk 1]
```diff
@@ -255,7 +255,7 @@ impl DataError {
     }
 
     #[inline]
-    pub(crate) fn for_type<T>() -> DataError {
+    pub fn for_type<T>() -> DataError {
```
Comment (sffc) [inline provider/core/src/error.rs:258]: Nit: If you make this public, please:

1. Add docs
2. Consider whether a name like `mismatched_type<T>()` would be better than `for_type<T>()`

File: provider/datagen/src/transform/cldr/time_zones/mod.rs [hunk 2]
```diff
@@ -45,105 +33,72 @@ macro_rules! impl_resource_provider {
...
-                                .map_err(|e| DataError::from(e).with_path_context(&bcp47_time_zone_path))?;
-                        let r = resource.keyword.u.time_zones.values;
-
-                        let mut data_guard = self.bcp47_tzid_data.write().expect("poison");
-                        for (bcp47_tzid, bcp47_tzid_data) in r.iter() {
-                            if let Some(alias) = &bcp47_tzid_data.alias {
-                                for data_value in alias.split(" ") {
-                                    data_guard.insert(data_value.to_string(), *bcp47_tzid);
-                                }
+                    let mut bcp47_tzids = LiteMap::new();
```
Comment (sffc) [inline provider/datagen/src/transform/cldr/time_zones/mod.rs:56]: Thought: Before, we cached the LiteMap here, but now we have to re-compute the LiteMap each time. Unsure if there is a measurable perf impact. It's probably fine. But if we need to, we could re-introduce the `RwLock` in this specific case.

File: provider/datagen/src/transform/cldr/plurals/mod.rs [hunk 3]
```diff
@@ -3,87 +3,48 @@
...
-                        .expect("poison")
-                        .get_or_insert(data.supplemental.plurals_type_ordinal);
-                }
-                #[allow(clippy::unwrap_used)]
-                // TODO(#1668) Clippy exceptions need docs or fixing.
-                self.ordinal_rules.read().expect("poison")
-            }
-            _ => return Err(DataError::custom("Unknown key for PluralRulesV1")),
-        })
+    fn get_rules_for(&self, key: ResourceKey) -> Result<&cldr_serde::plurals::Rules, DataError> {
```
Comment (sffc) [inline provider/datagen/src/transform/cldr/plurals/mod.rs:27]: Praise: This function is much nicer. We can do this now with the centralized cache.

File: provider/datagen/src/transform/cldr/cldr_serde/ca.rs [hunk 4]
```diff
@@ -105,7 +105,7 @@ symbols!(
     ["midnight", midnight, Option<Cow<'static, str>>],
 );
 
-#[derive(PartialEq, Debug, Deserialize)]
+#[derive(PartialEq, Debug, Deserialize, Clone)]
```
Comment (sffc) [inline provider/datagen/src/transform/cldr/cldr_serde/ca.rs:108]: Question: For the structs where you had to add `Clone` impls, did you check whether it was easy to convert the transformers to working on references?

File: provider/datagen/src/transform/collator/mod.rs [hunk 5]
```diff
@@ -312,3 +307,40 @@ collation_provider!(
...
 );
+
+use std::fs::{self, File};
+use std::io::BufReader;
+use std::io::Read;
+/// TODO move into cache
+use std::path::{Path, PathBuf};
+
+/// Helper function to open a file and return failures as a crate error.
+fn open_reader(path: &Path) -> Result<BufReader<File>, DataError> {
```
Comment (sffc) [inline provider/datagen/src/transform/collator/mod.rs:318]: Nit: I'd prefer if you kept these functions in `crate::transform::reader` until no one needs them anymore. Moving them around like this just means they may need to be moved again and again.

Comment (robertbastian) [reply provider/datagen/src/transform/collator/mod.rs]: This was still TODO, added the TOML cache here as well.

File: provider/datagen/src/bin/datagen.rs [hunk 6]
```diff
@@ -244,49 +239,44 @@ fn main() -> eyre::Result<()> {
         eyre::bail!("No keys selected");
     }
 
+    let cldr_locale = matches
```
Comment (sffc) [inline provider/datagen/src/bin/datagen.rs]: Nit: `cldr_locale_type` or something, because it's not a locale

**Review by sffc — APPROVED** [review]

File: provider/datagen/src/bin/datagen.rs (hunk 6 above)
Comment (robertbastian) [reply provider/datagen/src/bin/datagen.rs]: renaming in follow up

**Review by sffc — APPROVED** [review]

**Comment by robertbastian** [comment]
@Manishearth I'm merging because I won't get back to this for a few days and it will get conflicts. Please still take a look and I'll follow up

**Pull request merged**